crest/Assets/Crest/Crest/Scripts/Time/TimeProviderCutscene.cs

```diff
@@ -0,0 +1,119 @@
+    {
+        [Tooltip("Playable Director to take time from"), SerializeField]
+        PlayableDirector _playableDirector;
+
+        [Tooltip("Time offset which will be added to the Timeline time"), SerializeField]
+        float _timeOffset = 0f;
+
+        [Tooltip("Auto-assign this time provider to the ocean system when this component becomes active"), SerializeField]
+        bool _assignToOceanComponentOnEnable = true;
+
+        [Tooltip("Restore whatever time provider was previously assigned to ocean system when this component disables"), SerializeField]
+        bool _restorePreviousTimeProviderOnDisable = true;
+
+        TimeProviderDefault _fallbackTP = new TimeProviderDefault();
+
+        bool _initialised = false;
+
+        private void OnStart()
+        {
+            Initialise(true);
+        }
+
+        void Initialise(bool showErrors)
+        {
+            if (OceanRenderer.Instance == null)
+            {
+                if (showErrors)
+                {
+                    Debug.LogError("Crest: No ocean present, TimeProviderCutscene will have no effect.", this);
+                }
+#if !UNITY_EDITOR
+                enabled = false;
+#endif
+                return;
+            }
+
+            if (_playableDirector == null)
+            {
+                if (showErrors)
+                {
+                    Debug.LogError("Crest: No Playable Director component assigned, TimeProviderCutscene will have no effect.", this);
+                }
+#if !UNITY_EDITOR
+                enabled = false;
+#endif
+                return;
+            }
+
+            if (_assignToOceanComponentOnEnable)
+            {
+                OceanRenderer.Instance.PushTimeProvider(this);
+            }
+
+            _initialised = true;
+        }
+
+        private void OnDisable()
+        {
+            if (_restorePreviousTimeProviderOnDisable && OceanRenderer.Instance != null && _initialised)
+            {
+                OceanRenderer.Instance.PopTimeProvider(this);
+            }
+
+            _initialised = false;
+        }
+
+#if UNITY_EDITOR
+        // Needed to keep up to date while editing. Don't want to display errors when user
+        // is halfway through configuring component, and keep trying to initialise until
+        // everything is there.
+        private void Update()
+        {
+            if (!_initialised)
+            {
+                Initialise(false);
+            }
+        }
+#endif
+
+        // If there is a playable director which is playing, return its time, otherwise
+        // use whatever TP was being used before this component initialised, else fallback
+        // to a default TP.
+        public override float CurrentTime
+        {
+            get
+            {
+                if (_playableDirector != null
+                    && _playableDirector.isActiveAndEnabled
+                    && (!EditorApplication.isPlaying || _playableDirector.state == PlayState.Playing))
```

Comment:
Build error here. Could use _Application.isPlaying_ instead since it does the same thing when reading.

Reply:
did not know about Application.isPlaying, nice. done